drivers/clock_control/nrf_power_clock.c

```diff
@@ -9,272 +9,296 @@
+typedef void (*nrf_clock_stop_handler_t)(struct device *dev);
+
+/* Clock instance structure */
+struct nrf_clock_control {
+	s32_t ref;		/* Users counter */
+	sys_slist_t list;	/* List of users requesting callback */
+	bool started;		/* When set clock is started */
+};
 
-static int m16src_start(struct device *dev, clock_control_subsys_t sub_system)
-{
-	bool blocking;
-	u32_t imask;
-	u32_t stat;
+/* Clock instance static configuration */
+struct nrf_clock_control_config {
+	u32_t start_tsk;	/* Clock start task */
+	u32_t started_evt;	/* Clock started event */
+	u32_t stop_tsk;		/* Clock stop task */
+	nrf_clock_stop_handler_t stop_handler; /* Function called before stop */
+};
 
-	/* If the clock is already started then just increment refcount.
-	 * If the start and stop don't happen in pairs, a rollover will
-	 * be caught and in that case system should assert.
-	 */
+#define INT_MASK \
+	(NRF_CLOCK_INT_HF_STARTED_MASK | NRF_CLOCK_INT_LF_STARTED_MASK | \
+	COND_CODE_1(CONFIG_USB_NRF52840, \
+		    (NRF_POWER_INT_USBDETECTED_MASK | \
+		     NRF_POWER_INT_USBREMOVED_MASK | \
+		     NRF_POWER_INT_USBPWRRDY_MASK), (0)))
 
-	/* Test for reference increment from zero and resource guard not taken.
-	 */
-	imask = irq_lock();
+/* Return true if given event has enabled interrupt and is triggered. Event
+ * is cleared.
+ */
+static bool clock_event_check_and_clean(u32_t evt, u32_t intmask)
+{
+	bool ret = nrf_clock_event_check(evt) &&
+			nrf_clock_int_enable_check(intmask);
 
-	if (m16src_ref++) {
-		irq_unlock(imask);
-		goto hf_already_started;
+	if (ret) {
+		nrf_clock_event_clear(evt);
 	}
 
-	if (m16src_grd) {
-		m16src_ref--;
-		irq_unlock(imask);
-		return -EAGAIN;
-	}
+	return ret;
+}
+
+static enum clock_control_status get_status(struct device *dev,
+					    clock_control_subsys_t sys)
+{
+	struct nrf_clock_control *data = dev->driver_data;
 
-	m16src_grd = 1U;
+#if 0
+	__ASSERT_NO_MSG(z_arm_irq_current_prio() >
+			NVIC_GetPriority(nrfx_get_irq_number(NRF_CLOCK)));
+#endif
 
-	irq_unlock(imask);
+	if (data->ref > 0) {
+		return data->started ?
+			CLOCK_CONTROL_STATUS_ON : CLOCK_CONTROL_STATUS_STARTING;
+	}
 
-	/* If blocking then spin-wait in CPU sleep until 16MHz clock settles. */
-	blocking = POINTER_TO_UINT(sub_system);
-	if (blocking) {
-		u32_t intenset;
+	return CLOCK_CONTROL_STATUS_OFF;
+}
 
-		irq_disable(DT_INST_0_NORDIC_NRF_CLOCK_IRQ_0);
+static int clock_stop(struct device *dev, clock_control_subsys_t sub_system)
+{
+	const struct nrf_clock_control_config *config =
+						dev->config->config_info;
+	struct nrf_clock_control *data = dev->driver_data;
 
-		NRF_CLOCK->EVENTS_HFCLKSTARTED = 0;
+	int key = irq_lock();
+	int err = 0;
 
-		intenset = NRF_CLOCK->INTENSET;
-		nrf_clock_int_enable(NRF_CLOCK_INT_HF_STARTED_MASK);
+	data->ref--;
 
-		nrf_clock_task_trigger(NRF_CLOCK_TASK_HFCLKSTART);
+	if (data->ref == 0) {
+		DBG(dev, "Stopping");
+		sys_slist_init(&data->list);
```

Comment:
if clock was stopped before started then we need to clear the list. otherwise after starting again list will not be empty and unwanted callback may be called on next clock start.

Reply:
> if clock was stopped before started then we need to clear the list

Right. I haven't thought about this case. It's not that obvious, so perhaps a short comment here would be a good thing?